### DISCUSSION TITLE
`date: ` field, valid entries

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#4636] [Q&A]
I'm finding problems when converting old documents from Rmd to qmd, in that I tend to use custom formats for the date, such as:

- [``date: "`r format(Sys.time(), '%d %B, %Y')`"``](https://stackoverflow.com/questions/23449319/yaml-current-date-in-rmarkdown)
- `date: "\\the\\year, Trimester 1"`
- `date: 2023`
- ``date: !expr "`r format(Sys.time(), "%d %B, %Y")`"`` (or something like that, I can never remember/find the right syntax)

etc.

Quarto seems to perform some strict checks and either refuses to compile the document or, potentially worse, interprets the date as Y/M/D. 

Most of the time I just end up changing `date: ` with a custom variable such as `taught: Trimester 1` and use the corresponding variable in the custom pandoc template, but if I was using a default template that workaround would not be possible. Is there a way to relax the checks for yaml entries? Or some flexible formatting options I might have overlooked? (common use-cases as shown above include specifying the year only, with a custom string attached, etc.)

**Comment by dragonstyle** [comment]
In Quarto, we do our best to separate the date value and the formatting of the date. So, the best way to approach this, I think, is to specify a date and date-format separately. For example, to display only the year, you could write something like:

```markdown
---
title: Quarto Document
author: Norah Jones
date: 2022-10-13
date-format: YYYY
---

## Test

Lorem ipsum.
```

There are a few useful date macros that are treated specially (`today`, `now`, and `last-modified`). 

More about specifying dates and date formats here:

https://quarto.org/docs/reference/dates.html

HTH!

**Reply by dragonstyle** [reply]
Incidentally, we're generally heading towards allowing date values in Quarto to support specifying dates and formats together. For example:

```markdown
---
title: Quarto Document
author: Norah Jones
date: 
  value: 2022-10-13
  format: YYYY
---

## Test

Lorem ipsum.
```

(This form is currently support for `date` and `date-modified` values)

**Reply by mcanouil** [reply]
I am using in different places a "complicated" format, where date is only the ISO value for date.

```yaml
date-format: 'dddd[, the] D [of] MMMM, YYYY'
```

This avoid the need to have computation with `knitr` for example.

Note that the syntax for including code are (the second is preferred):
```yaml
date: "`r format(Sys.time(), '%d %B, %Y')`"
date: !expr "format(Sys.time(), '%d %B, %Y')"
```

**Reply by baptiste** [reply]
> More about specifying dates and date formats here:
> 
> https://quarto.org/docs/reference/dates.html

Thanks, that's very helpful – somehow I missed that page in my search.

It's great that the implementation is flexible enough to allow custom formats, though personally I would have gone the other way for the `date-format`, allowing normal text without escaping but having a special syntax for interpolating variables. Something like `Trimester 1, ${YYYY}`, as it can be quite counterintuitive to get an innocuous character like `a` transformed into `am/pm` etc.

**Reply by baptiste** [reply]
> Note that the syntax for including code are (the second is preferred):
> 
> ```yaml
> date: "`r format(Sys.time(), "%d %B, %Y")`"
> date: !expr "format(Sys.time(), '%d %B, %Y')"
> ```

Neither of those seems to work in Quarto though:

````
YAMLError: end of the stream or a document separator is expected at line 4, column 31:
    date: "`r format(Sys.time(), "%d %B, %Y")`"
                                  ^
````

and

````
Internal Error, couldn't locate key value
````

It might be worth putting a note somewhere on the date docs, as [this StackOverflow question](https://stackoverflow.com/q/23449319) has had a lot of views and many quarto users will likely go through the same exercise of trying (doomed to fail) syntax variations.

**Reply by baptiste** [reply]
.. That said, 

```
date: "`r format(Sys.time(), '%d %B, %Y')`"
```

_does work_ (fixing the single/double quote issue), but here's the unexpected bit: it produces `Invalid Date` in the document if the output value is not a valid date, as in for example,

```
date: "`r paste(format(Sys.time(), '%d %B, %Y'), 'Trimester 1')`"
```

which I think will lead to surprises.

**Reply by cscheid** [reply]
My (personal, 100% unofficial!) perspective for `r foo()` and `!expr foo()` in quarto metadata is: "this stuff will generally void your warranty and risk [nasal demons](https://en.wiktionary.org/wiki/nasal_demon)". We can't stop downstream packages from more or less doing it, but if you use that stuff, we can't promise much ourselves, and as time goes by, it becomes more likely that it will break your .qmd files in a hard-to-detect way.

**Reply by mcanouil** [reply]
@baptiste I copied pasted and did not fixed all the single/double quotes.

The syntax now is correct at least for code cells relying on knitr.

These syntaxes are 100 % `knitr` thus why @cscheid gently stated the "at your own risks" thing 😂
Note that in RMarkdown there is no date supports, thus we have to handle dates in R, the R way.
Quarto is agnostic.


For short, @baptiste either embrace Quarto syntax or stick to RMarkdown 😅

**Reply by baptiste** [reply]
> My (personal, 100% unofficial!) perspective for `r foo()` and `!expr foo()` in quarto metadata is: "this stuff will generally void your warranty and risk [nasal demons](https://en.wiktionary.org/wiki/nasal_demon)". We can't stop downstream packages from more or less doing it, but if you use that stuff, we can't promise much ourselves, and as time goes by, it becomes more likely that it will break your .qmd files in a hard-to-detect way.



That's fair, and I think it would be worth adding a note along those lines in the docs, under a FAQ for former Rmd users/hackers kind of section. 

That being said, I do think systems like Quarto will always hugely benefit from allowing custom syntax extensions (/ "hacks") that go beyond the original framework. I just came across [typst](https://typst.app/docs/tutorial/advanced-styling/), whose macro syntax reminds me a little of [Pollen](https://docs.racket-lang.org/pollen/programming-pollen.html), and of course LaTeX macros, and pandoc template variables, etc. What I'm trying to say is: allowing "arbitrary hacks" into the system is a huge part of what makes these tools successful, because they let users implement almost arbitrary new ideas (the "make it easy to do easy things, possible to do hard things" adage), some of which might later prove sufficiently valuable & desirable to be properly implemented and supported.

My personal view is that it would be nice to have a [generic pre-processor mechanism](https://github.com/quarto-dev/quarto-cli/discussions/3150) that _produces_ a valid and vanilla .qmd file for quarto to process, from an input file with (almost arbitrary) custom syntax extensions (provided the transformer know what to do – and I think custom pandoc writers might be a good example of that, or Pollen, actually).

**Reply by mcanouil** [reply]
Based on the original post items.

### From `` date: "`r format(Sys.time(), '%d %B, %Y')`" ``[^1]

```yml
date: today
date-format: DD MMMM, YYYY
```

Alternatively, there is an ugly way to do it with `knitr`:

````qmd
---
title: "My document"
format: html
---

```{r}
#| results: asis
#| echo: false
cat(
  "---\n",
  "date: ", format(Sys.time(), "%d %B, %Y"), "\n",
  "---\n",
  sep = ""
)
```
````

[^1]: Apparently, the new `!expr` prefix does not seem to be supported for date in RMarkdown as well.

### From `date: "\\the\\year, Trimester 1"`

```yml
date: today
date-format: YYYY[, Trimester 1]
```

### From `date: 2023`

```yml
date: today
date-format: YYYY
```

**Reply by baptiste** [reply]
> For short, @baptiste either embrace Quarto syntax or stick to RMarkdown 😅

I'm happily switching for most things (except when on a deadline, etc.), and I think it's fair to say that for many long-time users of knitr/Rmd (I've used it for over a decade) Quarto feels like it has learned and benefited from all the positive things that were developed in that ecosystem. But to me it also points to the huge benefit of having a flexible (hackable) system, not just to ensure that old hacks can still survive (that's not very important, as long as there is an equivalent, proper way to do things), but more importantly to develop _new_ ideas.

**Reply by mcanouil** [reply]
I just posted flexible hackable solutions from the formats you highlighted in your post.

**Reply by cscheid** [reply]
@baptiste I hear you. For the deep hackers like you, quarto is currently at an awkward spot where not everything that should be hackable is. We appreciate your patience and I hope you believe me when I say that I was an emacs user for 25 years of my life for good reason, and making quarto fundamentally hackable is a _central concern of mine_ (I can say that with 100% certainty for myself, and with 90% certainty for the other folks in the team :)).

The biggest challenge is how to make it both hackable and engine-agnostic. We're making progress, but there are a lot of parts where we could do better! We hope we can keep you deep hackers interested while we get everything ready <3
